soc/nordic/common/mram_latency.c

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <mram_latency.h>
+#include <zephyr/kernel.h>
+#include <nrfs_mram.h>
+#include <nrfs_backend_ipc_service.h>
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(mram_latency, CONFIG_MRAM_LATENCY_LOG_LEVEL);
+
+static struct k_work work;
+static bool no_latency;
+static bool init_ok;
+
+static onoff_notify_fn onoff_notify;
+struct onoff_manager mram_latency_mgr;
+
+static void latency_change_req(bool latency_not_allowed)
+{
+	nrfs_err_t err;
+
+	if (latency_not_allowed) {
+		err = nrfs_mram_set_latency(MRAM_LATENCY_NOT_ALLOWED, NULL);
+		if (err != NRFS_SUCCESS) {
+			onoff_notify(&mram_latency_mgr, -EIO);
+		}
+	} else {
+		/* There is no event for that setting so we can notify onoff manager
+		 * immediately.
+		 */
+		err = nrfs_mram_set_latency(MRAM_LATENCY_ALLOWED, NULL);
+		onoff_notify(&mram_latency_mgr, err != NRFS_SUCCESS ? -EIO : 0);
+	}
+}
+
+static void latency_change(bool latency_not_allowed)
+{
+	LOG_DBG("Request: latency %s allowed", latency_not_allowed ? "not " : "");
+	if (k_is_in_isr()) {
+		/* nrfs cannot be called from interrupt context so defer to the work
+		 * queue context and execute from there.
+		 */
+		no_latency = latency_not_allowed;
+		k_work_submit(&work);
+	} else {
+		latency_change_req(latency_not_allowed);
+	}
+}
+
+static void no_latency_start(struct onoff_manager *mgr, onoff_notify_fn notify)
+{
+	onoff_notify = notify;
+	latency_change(true);
+}
+
+static void no_latency_stop(struct onoff_manager *mgr, onoff_notify_fn notify)
+{
+	latency_change(false);
+}
+
+
+static void evt_handler(nrfs_mram_latency_evt_t const * p_evt, void * context)
+{
+	int res = p_evt->type == NRFS_MRAM_LATENCY_REQ_APPLIED ? 0 : -EIO;
+
+	LOG_DBG("Latency not allowed - applied");
+	onoff_notify(&mram_latency_mgr, res);
+}
+
+static void work_handler(struct k_work *work)
+{
+	latency_change_req(no_latency);
+}
+
+int mram_no_latency_cancel_or_release(struct onoff_client *cli)
+{
+	if (!init_ok) {
+		return -EIO;
+	}
+
+	return onoff_cancel_or_release(&mram_latency_mgr, cli);
+}
+
+int mram_no_latency_request(struct onoff_client *cli)
+{
+	if (!init_ok) {
+		return -EIO;
+	}
+
+	return onoff_request(&mram_latency_mgr, cli);
+}
+
+static int init_mram_latency(void)
+{
+	static const struct onoff_transitions transitions =
+		ONOFF_TRANSITIONS_INITIALIZER(no_latency_start, no_latency_stop, NULL);
+	nrfs_err_t err;
+	int rv;
+
+	err = nrfs_backend_wait_for_connection(K_FOREVER);
+	if (err != NRFS_SUCCESS) {
+		rv = -EIO;
+		goto err;
+	}
+
+	err = nrfs_mram_init(evt_handler);
+	if (err != NRFS_SUCCESS) {
+		rv = -EIO;
+		goto err;
+	}
+
+	k_work_init(&work, work_handler);
+	rv = onoff_manager_init(&mram_latency_mgr, &transitions);
+	init_ok = (rv == 0);
+err:
+	return rv;
+
+}
+
+SYS_INIT(init_mram_latency, APPLICATION, CONFIG_APPLICATION_INIT_PRIORITY);
```

Comment:
is only application expected to call this API?

Reply:
I've split the init into 2 stages, first onoff manager is initialized and then nrfs. Requests that comes before nrfs is initialized will not get completion notification until nrfs is initialized.